pairing/bn254/suite.go

```diff
@@ -0,0 +1,209 @@
+// package bn254 implements a particular bilinear group.
+//
+// Bilinear groups are the basis of many of the new cryptographic protocols that
+// have been proposed over the past decade. They consist of a triplet of groups
+// (G₁, G₂ and GT) such that there exists a function e(g₁ˣ,g₂ʸ)=gTˣʸ (where gₓ
+// is a generator of the respective group). That function is called a pairing
+// function.
+//
+// This package specifically implements the Optimal Ate pairing over a 256-bit
+// Barreto-Naehrig curve as described in
+// http://cryptojedi.org/papers/dclxvi-20100714.pdf. Its output is compatible
+// with the implementation described in that paper.
+//
+// This package previously claimed to operate at a 128-bit security level.
+// However, recent improvements in attacks mean that is no longer true. See
+// https://moderncrypto.org/mail-archive/curves/2016/000740.html.
+package bn254
+
+import (
+	"crypto/cipher"
+	"hash"
+	"io"
+	"reflect"
+
+	"github.com/drand/kyber"
+	"github.com/drand/kyber/util/random"
+	"github.com/drand/kyber/xof/blake2xb"
+	"go.dedis.ch/fixbuf"
+	"golang.org/x/crypto/sha3"
+)
+
+// Suite implements the pairing.Suite interface for the BN254 bilinear pairing.
+type Suite struct {
+	*commonSuite
+	g1 *groupG1
+	g2 *groupG2
+	gt *groupGT
+}
+
+var DEFAULT_DOMAIN_G1 = []byte("BLS_SIG_BN254G1_XMD:KECCAK-256_SSWU_RO_NUL_")
+var DEFAULT_DOMAIN_G2 = []byte("BLS_SIG_BN254G2_XMD:KECCAK-256_SSWU_RO_NUL_")
```

Comment:
This doesn't matches [RFC9380 naming convention](https://datatracker.ietf.org/doc/html/rfc9380#section-8.10):

> `CURVE_ID || "_" || HASH_ID || "_" || MAP_ID || "_" || ENC_VAR || "_"`
> The fields CURVE_ID, HASH_ID, MAP_ID, and ENC_VAR are ASCII-encoded strings of at most 64 characters each. Fields MUST contain only ASCII characters between 0x21 and 0x7E (inclusive), except that underscore (i.e., 0x5F) is not allowed.

So, I think this would be more correct:
CURVE_ID = BN254G1
HASH_ID = XMD:KECCAK-256
MAP_ID = SSWU (see other comment, but I think it checks out that this is SSWU for BN curves)
ENC_VAR=RO

So:
```suggestion
var DEFAULT_DOMAIN_G1 = []byte("BN254G1_XMD:KECCAK-256_SSWU_RO_")
var DEFAULT_DOMAIN_G2 = []byte("BN254G2_XMD:KECCAK-256_SSWU_RO_")
```

What is the `RO_NUL_` in the original proposal, why the extra `_NUL` ?

Reply:
I'm having a hard time convincing myself either way. It could be SVDW but it looks like Fouque-Tibouchi came after the first simplifications for SVDW were found which are SSWU, no?

Reply:
SSWU in RFC9380 seems to explicitly refer to the algorithm defined in [6.6.2](https://datatracker.ietf.org/doc/html/rfc9380#simple-swu). In our case, BN254 has curve A parameter == 0 so we actually need to additionally compute an isogenous curve as in [6.6.3](https://datatracker.ietf.org/doc/html/rfc9380#section-6.6.3). From reading [Wahby & Boneh](https://eprint.iacr.org/2019/403.pdf) pg2, it sounds like the Fouque-Tibouchi mapping is explicitly different to all the others? SvdW, Fouque-Tibouchi, and the SSWU mappings all output different points:

```
❯ sage suite_bn254.sage
u:  7105195380181880595384217009108718366423089053558315283835256316808390512725
svdw:
 15712026073284912390314437469998384224444098668487062629391055065992760594476 12286200326952730997678485294504458874299852441720220164574895986935631271221

ft:
 19485131671658523517646027848906165907640971588430452127920614621547697012573 7252485661626054658053752721536032361940074412825453078837989033903251969412

generic sswu:
 (7433244435151743403934667274157583038597013229141355912918907345679928483392 : 3341345691842296612745507125415299735564087771630588448932624272206506288268 : 1)

optimised sswu:
 (7433244435151743403934667274157583038597013229141355912918907345679928483392 : 3341345691842296612745507125415299735564087771630588448932624272206506288268 : 1)
```
Code here: https://github.com/kevincharm/draft-irtf-cfrg-hash-to-curve/pull/1/files

So I think it's neither SVDW or SSWU?

Reply:
This is great, thanks a lot @kevincharm, we now know we don't have an easy way out I guess!

Let me cheat: @armfazh, I see your name on the RFC, can you help us out here? 
How should we define our Suite ID for the curve BN254 if we're using Fouque-Tibouchi for hash to curve? 
Or in the frame of writing RFC9380, did you somehow conclude it HAD to be either SSWU or SVDW but not Fouque-Tibouchi and so you've decided to not offer an FT option for some reasons?

For BN254, since AB = 0 we need to use an isogenous curve to do SSWU sadly and that seems more expensive / less practical to do in Solidity. So it seems our two options are either SVDW or FT, with the latter being more optimised for BN curves.

Reply:
Turns out I could actually implement the general SVDW cheaper than Fouque-Tibouchi in Solidity. Thanks @AnomalRoil for your suggestion to even try it 😁 (here's the implementation: https://github.com/kevincharm/sol-rfc9380/blob/master/contracts/SVDW.sol#L38)

Shall we just switch to SVDW?

Reply:
Yeah, then let's use SVDW 👍🏻  Thanks a lot

Reply:
@AnomalRoil I've updated the code to use SVDW mapping, and tested it against vectors from sagemath scripts from the RFC (and also the new Solidity & ts implementation at https://github.com/kevincharm/bls-bn254).

Also replaced the `expand_message_xmd` implementation with a better one from kilic/bls12-381 (and tested it against the gnark one, both of these modified to use keccak256 instead of sha256).

Reply:
Thanks a lot! I'll review it next week at the earliest 🙈

Reply:
No worries, I really appreciate your time in reviewing this :)